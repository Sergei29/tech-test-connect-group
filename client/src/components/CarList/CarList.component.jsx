import React from "react";
import { useCarList } from "../../hooks/useCarList";
//components:
import CarListItem from "./components/CarListItem";
//style:
import { CarListContainer } from "./CarList.style";

/**
 * @description functional component for car list
 * @returns {JSX} component markup
 */
const CarList = () => {
  const { loading, carList, error } = useCarList();

  let displayCarList = null;
  if (loading) displayCarList = <p>Loading...</p>;
  if (error) displayCarList = <p>{error}</p>;
  if (carList.length) {
    displayCarList = carList.map(
      ({ id, media, url, modelYear, description, price }) => (
        <CarListItem
          key={id}
          carName={`Jaguar ${modelYear}`}
          imgUrl={`/${media[0].url}`}
          carDetailsUrl={url}
          carSummary={description}
          carPrice={price}
        />
      )
    );
  }

  return <CarListContainer>{displayCarList}</CarListContainer>;
};

export default CarList;
