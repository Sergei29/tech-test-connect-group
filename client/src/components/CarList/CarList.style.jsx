import styled from "styled-components";
import { mediaQueryTo } from "../../constants/styleMixins";

export const CarListContainer = styled.div`
  display: grid;
  grid-template-columns: 1fr 1fr 1fr 1fr 1fr;
  column-gap: 1px;
  div {
    a {
      &:hover {
        opacity: 0.8;
        transition: all 0.3s ease-in-out;
      }
      img {
        max-width: 100%;
      }
    }
  }
  ${mediaQueryTo.md`
    grid-template-columns: auto auto;
  `}
  ${mediaQueryTo.xs`
    grid-template-columns: auto;
  `}
`;
