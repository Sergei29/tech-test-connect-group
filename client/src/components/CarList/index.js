export { default } from "./CarList.component";
