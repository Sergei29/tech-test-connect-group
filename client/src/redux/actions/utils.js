/**
 * @description generates the error message out of received error from api
 * @param {Object} error error received from api
 * @param {String} entity name of the entity: 'carList' or 'carCurrent'
 * @returns {String} aggregated error message
 */
export const getErrorMessage = (error, entity) => {
  let errorMessage = "API call failed.";

  if (error.response && error.response.data) {
    errorMessage = error.response.data;
  } else if (error.message) {
    errorMessage = error.message;
  }

  return `${errorMessage}. ${entity}`;
};
