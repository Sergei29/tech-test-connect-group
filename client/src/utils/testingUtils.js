import React from "react";
import { shallow, mount } from "enzyme";
import { BrowserRouter } from "react-router-dom";
import { Provider } from "react-redux";
import configureMockStore from "redux-mock-store";
import thunk from "redux-thunk";

const mockStore = configureMockStore([thunk]);

const store = mockStore({
  carList: [
    {
      id: "xe",
      modelYear: "k17",
      url: "/vehicle/xe",
      description:
        "The most advanced, efficient and refined sports saloon that Jaguar has ever produced",
      price: "£30,000",
      media: [{ media: "vehicle", url: "assets/images/xe_k17.jpg" }],
    },
    {
      id: "xf",
      modelYear: "k17",
      url: "/vehicle/xf",
      description:
        "Luxury business saloon with distinctive design, dynamic drive and state-of-the-art technologies.",
      price: "£36,000",
      media: [{ media: "vehicle", url: "assets/images/xf_k17.jpg" }],
    },
  ],
  carCurrent: {
    id: "xe",
    meta: {
      passengers: 5,
      drivetrain: ["AWD", "RWD"],
      bodystyles: ["saloon"],
      emissions: { template: "CO2 Emissions $value g/km", value: 99 },
    },
  },
  error: null,
  loading: false,
});

export const shallowRendererReduxConnected = (ComponentName, props = {}) =>
  shallow(
    <Provider store={store}>
      <BrowserRouter>
        <ComponentName {...props} />
      </BrowserRouter>
    </Provider>
  );

export const mountRendererReduxConnected = (ComponentName, props = {}) =>
  mount(
    <Provider store={store}>
      <BrowserRouter>
        <ComponentName {...props} />
      </BrowserRouter>
    </Provider>
  );

export const shallowRenderer = (ComponentName, props = {}) =>
  shallow(<ComponentName {...props} />);

export const mountRenderer = (ComponentName, props = {}) =>
  mount(
    <BrowserRouter>
      <ComponentName {...props} />
    </BrowserRouter>
  );
