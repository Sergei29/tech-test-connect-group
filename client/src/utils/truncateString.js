export const truncateString = (stringValue, maxLength) => {
  let truncateLocation = stringValue.indexOf(" ", maxLength);
  if (truncateLocation === -1) return stringValue;
  return stringValue.substring(0, truncateLocation) + "...";
};
