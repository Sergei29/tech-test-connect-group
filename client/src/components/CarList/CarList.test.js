import { mountRendererReduxConnected } from "../../utils/testingUtils";
import CarList from "./CarList.component";

describe("CarList", () => {
  let mountWrapper;

  beforeEach(() => {
    mountWrapper = mountRendererReduxConnected(CarList);
  });

  it("should match the snapshot", () => {
    expect(mountWrapper).toMatchSnapshot();
  });

  it("should contain CarListItem component(s)", () => {
    expect(mountWrapper.find("CarListItem").length > 0).toBe(true);
  });
});
