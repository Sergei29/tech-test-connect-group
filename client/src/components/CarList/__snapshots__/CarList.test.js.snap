// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CarList should match the snapshot 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <BrowserRouter>
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "createHref": [Function],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <CarList>
        <styled.div>
          <div
            className="sc-gsTCUz gooFPB"
          >
            <CarListItem
              carDetailsUrl="/vehicle/xe"
              carName="Jaguar k17"
              carPrice="£30,000"
              carSummary="The most advanced, efficient and refined sports saloon that Jaguar has ever produced"
              imgUrl="/assets/images/xe_k17.jpg"
              key="xe"
            >
              <Styled(Link)
                to="/vehicle/xe"
              >
                <Link
                  className="sc-bdfBwQ KGYe"
                  to="/vehicle/xe"
                >
                  <LinkAnchor
                    className="sc-bdfBwQ KGYe"
                    href="/vehicle/xe"
                    navigate={[Function]}
                  >
                    <a
                      className="sc-bdfBwQ KGYe"
                      href="/vehicle/xe"
                      onClick={[Function]}
                    >
                      <div
                        className="listItem__image_box"
                      >
                        <img
                          alt="vehicle Jaguar k17"
                          src="/assets/images/xe_k17.jpg"
                        />
                      </div>
                      <h3
                        className="listItem__car_name"
                      >
                        <span>
                          Jaguar k17
                        </span>
                      </h3>
                      <p
                        className="listItem__car_price"
                      >
                        £30,000
                      </p>
                      <p
                        className="listItem__car_summary"
                      >
                        The most advanced, efficient and refined sports saloon...
                      </p>
                    </a>
                  </LinkAnchor>
                </Link>
              </Styled(Link)>
            </CarListItem>
            <CarListItem
              carDetailsUrl="/vehicle/xf"
              carName="Jaguar k17"
              carPrice="£36,000"
              carSummary="Luxury business saloon with distinctive design, dynamic drive and state-of-the-art technologies."
              imgUrl="/assets/images/xf_k17.jpg"
              key="xf"
            >
              <Styled(Link)
                to="/vehicle/xf"
              >
                <Link
                  className="sc-bdfBwQ KGYe"
                  to="/vehicle/xf"
                >
                  <LinkAnchor
                    className="sc-bdfBwQ KGYe"
                    href="/vehicle/xf"
                    navigate={[Function]}
                  >
                    <a
                      className="sc-bdfBwQ KGYe"
                      href="/vehicle/xf"
                      onClick={[Function]}
                    >
                      <div
                        className="listItem__image_box"
                      >
                        <img
                          alt="vehicle Jaguar k17"
                          src="/assets/images/xf_k17.jpg"
                        />
                      </div>
                      <h3
                        className="listItem__car_name"
                      >
                        <span>
                          Jaguar k17
                        </span>
                      </h3>
                      <p
                        className="listItem__car_price"
                      >
                        £36,000
                      </p>
                      <p
                        className="listItem__car_summary"
                      >
                        Luxury business saloon with distinctive design, dynamic...
                      </p>
                    </a>
                  </LinkAnchor>
                </Link>
              </Styled(Link)>
            </CarListItem>
          </div>
        </styled.div>
      </CarList>
    </Router>
  </BrowserRouter>
</Provider>
`;
